wal/wal.go

```diff
@@ -908,47 +909,54 @@ type LiveReader struct {
 	index      int   // Used to track partial records, should be 0 at the start of every new record.
 }
 
+// Err returns any errors encountered reading the WAL.  io.EOFs are not terminal
+// and Next can be tried again.  Non-EOFs are terminal, and the reader should
+// not be used again.  It is up to the user to decide when to stop trying should
+// io.EOF be returned.
 func (r *LiveReader) Err() error {
 	return r.err
 }
 
+// TotalRead returns the number of bytes consumed from this segment.
 func (r *LiveReader) TotalRead() int64 {
 	return r.total
 }
 
-func (r *LiveReader) fillBuffer() error {
+func (r *LiveReader) fillBuffer() (int, error) {
 	n, err := r.rdr.Read(r.buf[r.writeIndex:len(r.buf)])
 	r.writeIndex += n
-	return err
-}
-
-// Shift the buffer up to the read index.
-func (r *LiveReader) shiftBuffer() {
-	copied := copy(r.buf[0:], r.buf[r.readIndex:r.writeIndex])
-	r.readIndex = 0
-	r.writeIndex = copied
+	return n, err
 }
 
-// Next returns true if r.rec will contain a full record.
-// False does not indicate that there will never be more data to
-// read for the current io.Reader.
+// Next returns true if Record() will contain a full record.
+// False guarantees there are no more records if the segment is closed,
```

Comment:
this is true except in the case of potentially corrupted segments like the one we've been looking at (6141), but we're still undecided if that segment is the result of a bug